### ISSUE TITLE
How to modify a response

### ISSUE THREAD

**Issue by umaar** [puppeteer/puppeteer#599]
Is it possible to modify a response? For example the headers/response body?

```js
page.on('response', response => {
  // Is there a way to modify the response here?
});
```

I understand it's possible to modify the request, but being able to modify a response too would be useful!

**Comment by ebidel** [comment]
Mind if I ask about the use case? I suppose you could modify the response body once you have it or update some of the `Response` properties.

The interface is defined [here](https://github.com/GoogleChrome/puppeteer/blob/master/lib/NetworkManager.js#L286-L292).

**Comment by aslushnikov** [comment]
> Is it possible to modify a response? For example the headers/response body?

For now, interception kicks in only before the request is sent, so there's no way to intercept responses.

There are some vague discussions about adding another interception level right after the response was received. This would allow amending responses, but there are no exact plans or ETA.

**Comment by umaar** [comment]
@ebidel intercepting/rewriting responses with WebDriver + a proxy is something I covered in my [browser automation talk](https://youtu.be/EufdahOSJIc?t=31m18s) and it was well received. In that same video, I also discuss use cases. For example: maybe I'm running loads of automation tests which all together, are slow. The webpage I'm testing overlays some animated popover on each page load. Being able to dynamically rewrite the HTML and/or JavaScript response to remove the necessary chunks of code is useful.

By the way that's brittle for production-grade tests, but just trying to give an example of a use case!

@aslushnikov as a side note, in the DevTools protocol, I see [Network.continueInterceptedRequest](https://chromedevtools.github.io/devtools-protocol/tot/Network/#method-continueInterceptedRequest) accepts a `rawResponse` - this is close! But not exactly what I'm looking for.

If it's not possible, then fair enough, however I'd recommend considering it again as I think it's very useful and offers a powerful feature to developers which can be quite challenging to get set up via WebDriver.

**Comment by aslushnikov** [comment]
> I see Network.continueInterceptedRequest accepts a rawResponse - this is close! But not exactly what I'm looking for.

Yes, this impersonates responses rather then intercepts them.

> I think it's very useful and offers a powerful feature to developers which can be quite challenging to get set up via WebDriver.

cc @dgozman who was considering response interception.

**Comment by paulirish** [comment]
the protocol [supports providing a `rawResponse`](https://aslushnikov.github.io/vanilla-protocol-viewer/#Network.continueInterceptedRequest) to `continue`, although this isn't in puppeteer right now.

you'd have to already have the response, so this kinda means:

1) do request, get response, modify it. 
2) turn on interception. 
3) do request again, set `rawResponse`, continue.

little awkward, which is why there are now "vague discussions" :)

**Comment by umaar** [comment]
Thank you, at least it's possible to do!

**Comment by niieani** [comment]
Another use-case is relaxing Content Security Policy for testing.

**Comment by dotNetDR** [comment]
use `request` event and `respond` method
```
page.on('request', (request) => {
    if (request.url() === "http://abc.com/a.js") {
        request.respond({
            status: 200,
            contentType: 'application/javascript; charset=utf-8',
            body: 'console.log(1);'
        });
    } else {
        request.continue();
    }
});
```

**Comment by niieani** [comment]
My workaround to fix CSP in the comment here https://github.com/GoogleChrome/puppeteer/issues/1229#issuecomment-357469434.

**Comment by rodoabad** [comment]
@aslushnikov  @dotNetDR if I was to modify a JSON response...

I'd be doing something like this?

Assume `/user` returns a user object.

```js
        await page.setRequestInterception(true);
        page.on('request', (request) => {
            if (request.url() === `${process.env.URL_TO_TEST}/user`) {
                request.respond({
                    status: 200,
                    body: userResponse
                });
            } else {
                request.continue();
            }
        });

**Comment by rodoabad** [comment]
I wish there was an easier way to override responses in Puppeteer like `page.waitForResponse()` has the option to override.

**Comment by crobinson42** [comment]
Adding to the conversation - my use case is for server-side rendering content for emails. I have React applications that I bundle and put in my nodejs server directory for Puppeteer to render and capture some html. Here's me working solution: 

My demo file tree:

```
/assets
  - index.html
  - bundle.js
/render.js
```

```html
<!-- /assets/index.html -->
<html>
    <body>
        This is a test index.html file
        
        <div id="root"></div>
        
        <script src="/bundle.js"></script>
    </body>
</html>
```

```js
// /assets/bundle.js
(() => {
   document.getElementById('root').innerHTML = "The JS bundle loaded!! " + new Date().toISOString()
})()
```

```js
// /render.js

const puppeteer = require('puppeteer')
const fs = require('fs')
const path = require('path')

const browser = await puppeteer.launch();
const page = await browser.newPage();

const assetFilePath = path.join(__dirname, "./assets/index.html");

await page.setRequestInterception(true);

page.on("request", interceptedRequest => {
  const url = interceptedRequest.url();

  if (url.startsWith("file://") && !url.match(assetFilePath)) {
    interceptedRequest.respond({
      body: fs.readFileSync(
        path.join(__dirname, "./assets/", url.replace("file://", ""))
      )
    });
  } else {
    interceptedRequest.continue();
  }
});

await page.goto(`file://${assetFilePath}`, {
  waitUntil: "networkidle0"
});

const html = await page.evaluate(() => {
  // eslint-disable-next-line
  return document.documentElement.innerHTML;
});

console.log(html);

```

Anyone should be able to run this example by creating the files and copying the  ☝🏼content - install puppeteer `npm i pupeteer` then run the `render.js` with node `node render.js`

**Comment by crobinson42** [comment]
https://stackoverflow.com/questions/56794245/server-side-render-javascript-app-with-puppeteer/56794246#56794246

**Comment by cnt1992** [comment]
look forward to have the same feature too.

**Comment by darkguy2008** [comment]
Same here, I'd like to alter the received HTML response for dynamic Angular testing use-cases.

**Comment by evanrolfe** [comment]
+1 for this feature. I would really like to be able do intercept responses in exactly the same way that requests can be intercepted and then continued or aborted i.e.
```js
page.on("response", response => {
  response.body = "...";
  response.continue();
})
```

My use-case is for a "proxy-like" software which tracks network requests made by the browser using puppeteer, and lets you modify the requests in order to do penetration testing on a web app. Being able to modify the response sent back from the server is vital to this application.

**Comment by naycho334** [comment]
When you match the request that you want to edit its body, just get the URL and make a call using "node-fetch" or "request" modules, when you receive the body edit it then sends it as a response to the original request.

for example:

```
const requestModule = require("request");
const cheerio = require("cheerio");

page.on("request", async (request) => {
  // Match the url that you want
  const isMatched = /page-12/.test(request.url());

  if (isMatched) {
    // Make a new call
    requestModule({
      url: request.url(),
      resolveWithFullResponse: true,
    })
      .then((response) => {
        const { body, headers, statusCode, statusMessage } = response;
        const contentType = headers["content-type"];

        // Edit body using cheerio module
        const $ = cheerio.load(body);
        $("a").each(function () {
          $(this).attr("href", "/fake_pathname");
        });

        // Send response
        request.respond({
          ok: statusMessage === "OK",
          status: statusCode,
          contentType,
          body: $.html(),
        });
      })
      .catch(() => request.continue());
  } else request.continue();
});
```

**Comment by Nodirbek3d** [comment]
I am so in need of this feature, is it available yet? I need to get response from server, modify it and continue to page. Don't ask me why :)

**Comment by ghost** [comment]
A possible way to do it via Devtools protocol with Puppeteer (Using Fetch API):
Sample code: https://github.com/ChromeDevTools/devtools-protocol/issues/199#issuecomment-606953626

And a blog post about using Puppeteer to modify responses on the fly: https://medium.com/@jsoverson/using-chrome-devtools-protocol-with-puppeteer-737a1300bac0

Here is also a Chrome extension to remove a CSP meta tag from a document's response body: https://github.com/Sentero-esp12/Debugger-API-Fetch-example-Chrome-Extension (Using Debugger and Fetch APIs)

And the API methods: https://chromedevtools.github.io/devtools-protocol/tot/Fetch/
The order I used is: Fetch.enable -> on Fetch.requestPaused event -> Fetch.getResponseBody -> Fetch.fulfillRequest -> Fetch.disable

**Comment by kwasimensah** [comment]
My use case is I'm testing a page that uses javascript to do a post to backend and then changes window.location based on the response. By the time waitForResponse is fired the navigation has already happened and the response body for the API call is gone.

**Comment by leveryd** [comment]
+1 need this feature

**Comment by bjankord** [comment]
Sometimes it is desired just to change a specific aspect of the response.
e.g. change a value, add a value, or remove a value.

For this, currently we have to re-build ALL of what is in the response when intercepting the request. If this response will be changed later, the mock response will have to be changed as well, or will include outdated stuff.
And the mock response will be very verbose, even if it is just needed to change one entry out of hundreds.

I'd much rather just get the original API response, modify it for the values I need to test, and then send it along its way.

Edit: Cypress has [just recently added experimental support for these feature](https://docs.cypress.io/guides/references/changelog.html#5-1-0). Here is how they handle intercepting and modifying a response before passing it along.

```js
// Function
// You can supply a callback function which receives the request as the first argument
cy.route2('/users/**', (req) => {
  req.headers.accept = 'application/json'

  // To modify a JSON response,
  // it will be provided as a string
  // in req.body and must be
  // parsed and stringified properly.
  const requestBody = JSON.parse(req.body)

  req.body = JSON.stringify({
    ...requestBody,
    note: 'Custom note'
  })
})
```

**Comment by abacaj** [comment]
TLDR: use https://chromedevtools.github.io/devtools-protocol/tot/Fetch/

**Comment by dgozman** [comment]
[See documentation](https://playwright.dev/docs/network#modify-responses) for the official way to modify responses.

**Comment by niieani** [comment]
I've made a library to do this easily: https://github.com/niieani/puppeteer-intercept-and-modify-requests

**Comment by shotforsky** [comment]
> I've made a library to do this easily: https://github.com/niieani/puppeteer-intercept-and-modify-requests

Works great, thanks for sharing,  niieani.

**Comment by lwintermelon** [comment]
Oh man please take a look at playwright's page.route api, simple and elegant.
```
  await page.route('*/**/api/v1/fruits', async route => {
    const response = await route.fetch();
    const json = await response.json();
    json.push({ name: 'Loquat', id: 100 });
    // Fulfill using the original response, while patching the response body
    // with the given JSON object.
    await route.fulfill({ response, json });
  });
```

In puppeteer, the page.on("request") api and the docs for [Request Interception](https://pptr.dev/guides/network-interception#cooperative-intercept-mode) is messy, what's `Cooperative Intercept`? what's `priority `?  what's `page.setRequestInterception(true)`? Why it affects page.on("request") ？？？Messing up all things. To be honest, I don't really get it after reading it several times.

So, just let `page.on("request")` be `page.on("request")`，ok? Don't push so much things on it, give us a clean api. Intercept by url path is clean and simple and useful, and the underlying CDP `Fetch.enable` works this way. As for api compatibility, I don't know.

Add this feature, please.

For how to modify a response, see the answer above.

> And the API methods: https://chromedevtools.github.io/devtools-protocol/tot/Fetch/
> The order I used is: Fetch.enable -> on Fetch.requestPaused event -> Fetch.getResponseBody -> Fetch.fulfillRequest -> Fetch.disable
